Using goccy/go-json instead of encoding/json fails a few tests in an existing project

This draft PR https://github.com/swaggest/rest/pull/24 migrates to `goccy/go-json` by changing imports. Unfortunately such a change leads to test failures and a panic. 

Relevant GitHub Actions log: https://github.com/swaggest/rest/pull/24/checks?check_run_id=1861642353.

<details>
<summary>Panic message.</summary>

```
panic: runtime error: index out of range [824626223968] with length 730624 [recovered]
	panic: runtime error: index out of range [824626223968] with length 730624

goroutine 28 [running]:
testing.tRunner.func1.1(0xe85180, 0xc0006c8a80)
	/opt/hostedtoolcache/go/1.15.7/x64/src/testing/testing.go:1072 +0x46a
testing.tRunner.func1(0xc001370000)
	/opt/hostedtoolcache/go/1.15.7/x64/src/testing/testing.go:1075 +0x636
panic(0xe85180, 0xc0006c8a80)
	/opt/hostedtoolcache/go/1.15.7/x64/src/runtime/panic.go:975 +0x47a
github.com/goccy/go-json.(*Decoder).compileToGetDecoder(0xc001265278, 0xc0006e2d00, 0xc0006e2d00, 0x0, 0x0, 0x440191, 0x417ed0)
	/home/runner/work/rest/rest/vendor/github.com/goccy/go-json/decode_compile_race.go:16 +0x412
github.com/goccy/go-json.(*Decoder).decode(0xc001265278, 0xc0006c8a60, 0x14, 0x14, 0xc0000e7238, 0x4837e0, 0x4837e0)
	/home/runner/work/rest/rest/vendor/github.com/goccy/go-json/decode.go:71 +0x125
github.com/goccy/go-json.(*Decoder).decodeForUnmarshal(0xc001265278, 0xc0006c8a60, 0x14, 0x14, 0xc0006e2d00, 0xc00136a360, 0xc0006e2d00, 0xc00136a360)
	/home/runner/work/rest/rest/vendor/github.com/goccy/go-json/decode.go:84 +0x9b
github.com/goccy/go-json.Unmarshal(0xc0012652f8, 0x13, 0x20, 0xc0006e2d00, 0xc00136a360, 0x20, 0x4345f8)
	/home/runner/work/rest/rest/vendor/github.com/goccy/go-json/json.go:268 +0xe5
github.com/swaggest/rest/request.(*DecoderFactory).jsonParams.func1.1(0xc0006c8a40, 0x13, 0xc0012658b8, 0xc00134f8e8, 0xe4fe01, 0x101)
	/home/runner/work/rest/rest/request/factory.go:164 +0x3bb
github.com/swaggest/form.(*decoder).setFieldByType(0xc0006da900, 0xe4fee0, 0xc00136a300, 0x199, 0xc0006bee40, 0x6, 0x40, 0x0, 0x5)
	/home/runner/work/rest/rest/vendor/github.com/swaggest/form/decoder.go:194 +0x79c2
github.com/swaggest/form.(*decoder).traverseStruct(0xc0006da900, 0xe48660, 0xc00136a300, 0x199, 0x1143c20, 0xe48660, 0xc0006bee40, 0x0, 0x40, 0x0)
	/home/runner/work/rest/rest/vendor/github.com/swaggest/form/decoder.go:175 +0x385
github.com/swaggest/form.(*Decoder).Decode(0xc0006e2c80, 0xe0e920, 0xc00136a300, 0xc00130ff50, 0x0, 0x0, 0x0, 0x418620, 0xc0006b1cc0)
	/home/runner/work/rest/rest/vendor/github.com/swaggest/form/form_decoder.go:166 +0x385
github.com/swaggest/rest/request.makeDecoder.func1(0xc0006b9d00, 0xe0e920, 0xc00136a300, 0x0, 0x0, 0x5a191e, 0xc001370000)
	/home/runner/work/rest/rest/request/decoder.go:40 +0x213
github.com/swaggest/rest/request.(*decoder).Decode(0xc00130fe60, 0xc0006b9d00, 0xe0e920, 0xc00136a300, 0x0, 0x0, 0x0, 0x0)
	/home/runner/work/rest/rest/request/decoder.go:55 +0x14b
github.com/swaggest/rest/request_test.TestDecoder_Decode_jsonParam(0xc001370000)
	/home/runner/work/rest/rest/request/decoder_test.go:305 +0x179
testing.tRunner(0xc001370000, 0xef5d48)
	/opt/hostedtoolcache/go/1.15.7/x64/src/testing/testing.go:1123 +0x203
created by testing.(*T).Run
	/opt/hostedtoolcache/go/1.15.7/x64/src/testing/testing.go:1168 +0x5bc
```

</details>

--- a/codec.go
+++ b/codec.go
@@ -12,14 +12,12 @@ const (
 )
 
 var (
-	cachedOpcodeSets       []*opcodeSet
-	cachedOpcodeMap        unsafe.Pointer // map[uintptr]*opcodeSet
-	existsCachedOpcodeSets bool
-	cachedDecoder          []decoder
-	cachedDecoderMap       unsafe.Pointer // map[uintptr]decoder
-	existsCachedDecoder    bool
-	baseTypeAddr           uintptr
-	maxTypeAddr            uintptr
+	cachedOpcodeSets []*opcodeSet
+	cachedOpcodeMap  unsafe.Pointer // map[uintptr]*opcodeSet
+	cachedDecoder    []decoder
+	cachedDecoderMap unsafe.Pointer // map[uintptr]decoder
+	baseTypeAddr     uintptr
+	maxTypeAddr      uintptr
 )
 
 //go:linkname typelinks reflect.typelinks
@@ -69,7 +67,6 @@ func setupCodec() error {
 		return fmt.Errorf("too big address range %d", addrRange)
 	}
 	cachedOpcodeSets = make([]*opcodeSet, addrRange)
-	existsCachedOpcodeSets = true
 	cachedDecoder = make([]decoder, addrRange)
 	baseTypeAddr = min
 	maxTypeAddr = max

--- a/codec.go
+++ b/codec.go
@@ -71,7 +71,6 @@ func setupCodec() error {
 	cachedOpcodeSets = make([]*opcodeSet, addrRange)
 	existsCachedOpcodeSets = true
 	cachedDecoder = make([]decoder, addrRange)
-	existsCachedDecoder = true
 	baseTypeAddr = min
 	maxTypeAddr = max
 	return nil

--- a/codec.go
+++ b/codec.go
@@ -19,6 +19,7 @@ var (
 	cachedDecoderMap       unsafe.Pointer // map[uintptr]decoder
 	existsCachedDecoder    bool
 	baseTypeAddr           uintptr
+	maxTypeAddr            uintptr
 )
 
 //go:linkname typelinks reflect.typelinks
@@ -72,6 +73,7 @@ func setupCodec() error {
 	cachedDecoder = make([]decoder, addrRange)
 	existsCachedDecoder = true
 	baseTypeAddr = min
+	maxTypeAddr = max
 	return nil
 }
 

--- a/decode_compile.go
+++ b/decode_compile.go
@@ -245,7 +245,7 @@ func (d *Decoder) compileMap(typ *rtype, structName, fieldName string) (decoder,
 }
 
 func (d *Decoder) compileInterface(typ *rtype, structName, fieldName string) (decoder, error) {
-	return newInterfaceDecoder(typ, structName, fieldName), nil
+	return newInterfaceDecoder(d, typ, structName, fieldName), nil
 }
 
 func (d *Decoder) removeConflictFields(fieldMap map[string]*structFieldSet, conflictedMap map[string]struct{}, dec *structDecoder, baseOffset uintptr) {

--- a/decode_compile_norace.go
+++ b/decode_compile_norace.go
@@ -3,7 +3,7 @@
 package json
 
 func (d *Decoder) compileToGetDecoder(typeptr uintptr, typ *rtype) (decoder, error) {
-	if !existsCachedDecoder {
+	if typeptr > maxTypeAddr {
 		return d.compileToGetDecoderSlowPath(typeptr, typ)
 	}
 

--- a/decode_compile_race.go
+++ b/decode_compile_race.go
@@ -2,12 +2,14 @@
 
 package json
 
-import "sync"
+import (
+	"sync"
+)
 
 var decMu sync.RWMutex
 
 func (d *Decoder) compileToGetDecoder(typeptr uintptr, typ *rtype) (decoder, error) {
-	if !existsCachedDecoder {
+	if typeptr > maxTypeAddr {
 		return d.compileToGetDecoderSlowPath(typeptr, typ)
 	}
 

--- a/decode_interface.go
+++ b/decode_interface.go
@@ -10,13 +10,15 @@ type interfaceDecoder struct {
 	typ        *rtype
 	structName string
 	fieldName  string
+	dec        *Decoder
 }
 
-func newInterfaceDecoder(typ *rtype, structName, fieldName string) *interfaceDecoder {
+func newInterfaceDecoder(dec *Decoder, typ *rtype, structName, fieldName string) *interfaceDecoder {
 	return &interfaceDecoder{
 		typ:        typ,
 		structName: structName,
 		fieldName:  fieldName,
+		dec:        dec,
 	}
 }
 
@@ -70,26 +72,11 @@ func decodeWithTextUnmarshaler(s *stream, unmarshaler encoding.TextUnmarshaler) 
 	return nil
 }
 
-func (d *interfaceDecoder) decodeStream(s *stream, p unsafe.Pointer) error {
+func (d *interfaceDecoder) decodeEmptyInterface(s *stream, p unsafe.Pointer) error {
 	s.skipWhiteSpace()
 	for {
 		switch s.char() {
 		case '{':
-			runtimeInterfaceValue := *(*interface{})(unsafe.Pointer(&interfaceHeader{
-				typ: d.typ,
-				ptr: p,
-			}))
-			rv := reflect.ValueOf(runtimeInterfaceValue)
-			if rv.NumMethod() > 0 && rv.CanInterface() {
-				if u, ok := rv.Interface().(Unmarshaler); ok {
-					return decodeWithUnmarshaler(s, u)
-				}
-				if u, ok := rv.Interface().(encoding.TextUnmarshaler); ok {
-					return decodeWithTextUnmarshaler(s, u)
-				}
-				return nil
-			}
-			// empty interface
 			var v map[string]interface{}
 			ptr := unsafe.Pointer(&v)
 			if err := newMapDecoder(
@@ -97,7 +84,7 @@ func (d *interfaceDecoder) decodeStream(s *stream, p unsafe.Pointer) error {
 				stringType,
 				newStringDecoder(d.structName, d.fieldName),
 				interfaceMapType.Elem(),
-				newInterfaceDecoder(d.typ, d.structName, d.fieldName),
+				newInterfaceDecoder(d.dec, d.typ, d.structName, d.fieldName),
 				d.structName,
 				d.fieldName,
 			).decodeStream(s, ptr); err != nil {
@@ -109,7 +96,7 @@ func (d *interfaceDecoder) decodeStream(s *stream, p unsafe.Pointer) error {
 			var v []interface{}
 			ptr := unsafe.Pointer(&v)
 			if err := newSliceDecoder(
-				newInterfaceDecoder(d.typ, d.structName, d.fieldName),
+				newInterfaceDecoder(d.dec, d.typ, d.structName, d.fieldName),
 				d.typ,
 				d.typ.Size(),
 				d.structName,
@@ -171,6 +158,35 @@ func (d *interfaceDecoder) decodeStream(s *stream, p unsafe.Pointer) error {
 	return errNotAtBeginningOfValue(s.totalOffset())
 }
 
+func (d *interfaceDecoder) decodeStream(s *stream, p unsafe.Pointer) error {
+	runtimeInterfaceValue := *(*interface{})(unsafe.Pointer(&interfaceHeader{
+		typ: d.typ,
+		ptr: p,
+	}))
+	rv := reflect.ValueOf(runtimeInterfaceValue)
+	if rv.NumMethod() > 0 && rv.CanInterface() {
+		if u, ok := rv.Interface().(Unmarshaler); ok {
+			return decodeWithUnmarshaler(s, u)
+		}
+		if u, ok := rv.Interface().(encoding.TextUnmarshaler); ok {
+			return decodeWithTextUnmarshaler(s, u)
+		}
+		return nil
+	}
+	iface := rv.Interface()
+	ifaceHeader := (*interfaceHeader)(unsafe.Pointer(&iface))
+	typ := ifaceHeader.typ
+	if d.typ == typ || typ == nil {
+		// concrete type is empty interface
+		return d.decodeEmptyInterface(s, p)
+	}
+	decoder, err := d.dec.compileToGetDecoder(uintptr(unsafe.Pointer(typ)), typ)
+	if err != nil {
+		return err
+	}
+	return decoder.decodeStream(s, ifaceHeader.ptr)
+}
+
 func (d *interfaceDecoder) decode(buf []byte, cursor int64, p unsafe.Pointer) (int64, error) {
 	cursor = skipWhiteSpace(buf, cursor)
 	switch buf[cursor] {
@@ -182,7 +198,7 @@ func (d *interfaceDecoder) decode(buf []byte, cursor int64, p unsafe.Pointer) (i
 			stringType,
 			newStringDecoder(d.structName, d.fieldName),
 			interfaceMapType.Elem(),
-			newInterfaceDecoder(d.typ, d.structName, d.fieldName),
+			newInterfaceDecoder(d.dec, d.typ, d.structName, d.fieldName),
 			d.structName, d.fieldName,
 		)
 		cursor, err := dec.decode(buf, cursor, ptr)
@@ -195,7 +211,7 @@ func (d *interfaceDecoder) decode(buf []byte, cursor int64, p unsafe.Pointer) (i
 		var v []interface{}
 		ptr := unsafe.Pointer(&v)
 		dec := newSliceDecoder(
-			newInterfaceDecoder(d.typ, d.structName, d.fieldName),
+			newInterfaceDecoder(d.dec, d.typ, d.structName, d.fieldName),
 			d.typ,
 			d.typ.Size(),
 			d.structName, d.fieldName,

--- a/decode_interface.go
+++ b/decode_interface.go
@@ -42,7 +42,7 @@ var (
 	)
 )
 
-func decodeWithUnmarshaler(s *stream, unmarshaler Unmarshaler) error {
+func decodeStreamUnmarshaler(s *stream, unmarshaler Unmarshaler) error {
 	start := s.cursor
 	if err := s.skipValue(); err != nil {
 		return err
@@ -57,7 +57,7 @@ func decodeWithUnmarshaler(s *stream, unmarshaler Unmarshaler) error {
 	return nil
 }
 
-func decodeWithTextUnmarshaler(s *stream, unmarshaler encoding.TextUnmarshaler) error {
+func decodeStreamTextUnmarshaler(s *stream, unmarshaler encoding.TextUnmarshaler) error {
 	start := s.cursor
 	if err := s.skipValue(); err != nil {
 		return err
@@ -72,7 +72,7 @@ func decodeWithTextUnmarshaler(s *stream, unmarshaler encoding.TextUnmarshaler) 
 	return nil
 }
 
-func (d *interfaceDecoder) decodeEmptyInterface(s *stream, p unsafe.Pointer) error {
+func (d *interfaceDecoder) decodeStreamEmptyInterface(s *stream, p unsafe.Pointer) error {
 	s.skipWhiteSpace()
 	for {
 		switch s.char() {
@@ -166,10 +166,10 @@ func (d *interfaceDecoder) decodeStream(s *stream, p unsafe.Pointer) error {
 	rv := reflect.ValueOf(runtimeInterfaceValue)
 	if rv.NumMethod() > 0 && rv.CanInterface() {
 		if u, ok := rv.Interface().(Unmarshaler); ok {
-			return decodeWithUnmarshaler(s, u)
+			return decodeStreamUnmarshaler(s, u)
 		}
 		if u, ok := rv.Interface().(encoding.TextUnmarshaler); ok {
-			return decodeWithTextUnmarshaler(s, u)
+			return decodeStreamTextUnmarshaler(s, u)
 		}
 		return nil
 	}
@@ -178,7 +178,17 @@ func (d *interfaceDecoder) decodeStream(s *stream, p unsafe.Pointer) error {
 	typ := ifaceHeader.typ
 	if d.typ == typ || typ == nil {
 		// concrete type is empty interface
-		return d.decodeEmptyInterface(s, p)
+		return d.decodeStreamEmptyInterface(s, p)
+	}
+	if typ.Kind() == reflect.Ptr && typ.Elem() == d.typ || typ.Kind() != reflect.Ptr {
+		return d.decodeStreamEmptyInterface(s, p)
+	}
+	if s.char() == 'n' {
+		if err := nullBytes(s); err != nil {
+			return err
+		}
+		*(*interface{})(p) = nil
+		return nil
 	}
 	decoder, err := d.dec.compileToGetDecoder(uintptr(unsafe.Pointer(typ)), typ)
 	if err != nil {
@@ -188,6 +198,46 @@ func (d *interfaceDecoder) decodeStream(s *stream, p unsafe.Pointer) error {
 }
 
 func (d *interfaceDecoder) decode(buf []byte, cursor int64, p unsafe.Pointer) (int64, error) {
+	runtimeInterfaceValue := *(*interface{})(unsafe.Pointer(&interfaceHeader{
+		typ: d.typ,
+		ptr: p,
+	}))
+	rv := reflect.ValueOf(runtimeInterfaceValue)
+	iface := rv.Interface()
+	ifaceHeader := (*interfaceHeader)(unsafe.Pointer(&iface))
+	typ := ifaceHeader.typ
+	if d.typ == typ || typ == nil {
+		// concrete type is empty interface
+		return d.decodeEmptyInterface(buf, cursor, p)
+	}
+	if typ.Kind() == reflect.Ptr && typ.Elem() == d.typ || typ.Kind() != reflect.Ptr {
+		return d.decodeEmptyInterface(buf, cursor, p)
+	}
+	if buf[cursor] == 'n' {
+		if cursor+3 >= int64(len(buf)) {
+			return 0, errUnexpectedEndOfJSON("null", cursor)
+		}
+		if buf[cursor+1] != 'u' {
+			return 0, errInvalidCharacter(buf[cursor+1], "null", cursor)
+		}
+		if buf[cursor+2] != 'l' {
+			return 0, errInvalidCharacter(buf[cursor+2], "null", cursor)
+		}
+		if buf[cursor+3] != 'l' {
+			return 0, errInvalidCharacter(buf[cursor+3], "null", cursor)
+		}
+		cursor += 4
+		**(**interface{})(unsafe.Pointer(&p)) = nil
+		return cursor, nil
+	}
+	decoder, err := d.dec.compileToGetDecoder(uintptr(unsafe.Pointer(typ)), typ)
+	if err != nil {
+		return 0, err
+	}
+	return decoder.decode(buf, cursor, ifaceHeader.ptr)
+}
+
+func (d *interfaceDecoder) decodeEmptyInterface(buf []byte, cursor int64, p unsafe.Pointer) (int64, error) {
 	cursor = skipWhiteSpace(buf, cursor)
 	switch buf[cursor] {
 	case '{':

--- a/decode_ptr.go
+++ b/decode_ptr.go
@@ -68,7 +68,9 @@ func (d *ptrDecoder) decode(buf []byte, cursor int64, p unsafe.Pointer) (int64, 
 		if buf[cursor+3] != 'l' {
 			return 0, errInvalidCharacter(buf[cursor+3], "null", cursor)
 		}
-		*(*unsafe.Pointer)(p) = nil
+		if p != nil {
+			*(*unsafe.Pointer)(p) = nil
+		}
 		cursor += 4
 		return cursor, nil
 	}

--- a/decode_slice.go
+++ b/decode_slice.go
@@ -1,6 +1,7 @@
 package json
 
 import (
+	"reflect"
 	"sync"
 	"unsafe"
 )
@@ -61,6 +62,16 @@ func copySlice(elemType *rtype, dst, src sliceHeader) int
 
 //go:linkname newArray reflect.unsafe_NewArray
 func newArray(*rtype, int) unsafe.Pointer
+
+func (d *sliceDecoder) errNumber(buf []byte, offset int64) *UnmarshalTypeError {
+	return &UnmarshalTypeError{
+		Value:  "number",
+		Type:   reflect.SliceOf(rtype2type(d.elemType)),
+		Struct: d.structName,
+		Field:  d.fieldName,
+		Offset: offset,
+	}
+}
 
 func (d *sliceDecoder) decodeStream(s *stream, p unsafe.Pointer) error {
 	for {
@@ -140,10 +151,14 @@ func (d *sliceDecoder) decodeStream(s *stream, p unsafe.Pointer) error {
 				}
 				s.cursor++
 			}
+		case '-', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
+			return d.errNumber([]byte{s.char()}, s.totalOffset())
 		case nul:
 			if s.read() {
 				continue
 			}
+			goto ERROR
+		default:
 			goto ERROR
 		}
 	}

--- a/decode_slice.go
+++ b/decode_slice.go
@@ -248,7 +248,12 @@ func (d *sliceDecoder) decode(buf []byte, cursor int64, p unsafe.Pointer) (int64
 				}
 				cursor++
 			}
+		case '-', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
+			return 0, d.errNumber([]byte{buf[cursor]}, cursor)
+		default:
+			goto ERROR
 		}
 	}
+ERROR:
 	return 0, errUnexpectedEndOfJSON("slice", cursor)
 }

--- a/decode_slice.go
+++ b/decode_slice.go
@@ -63,7 +63,7 @@ func copySlice(elemType *rtype, dst, src sliceHeader) int
 //go:linkname newArray reflect.unsafe_NewArray
 func newArray(*rtype, int) unsafe.Pointer
 
-func (d *sliceDecoder) errNumber(buf []byte, offset int64) *UnmarshalTypeError {
+func (d *sliceDecoder) errNumber(offset int64) *UnmarshalTypeError {
 	return &UnmarshalTypeError{
 		Value:  "number",
 		Type:   reflect.SliceOf(rtype2type(d.elemType)),
@@ -152,7 +152,7 @@ func (d *sliceDecoder) decodeStream(s *stream, p unsafe.Pointer) error {
 				s.cursor++
 			}
 		case '-', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
-			return d.errNumber([]byte{s.char()}, s.totalOffset())
+			return d.errNumber(s.totalOffset())
 		case nul:
 			if s.read() {
 				continue
@@ -249,7 +249,7 @@ func (d *sliceDecoder) decode(buf []byte, cursor int64, p unsafe.Pointer) (int64
 				cursor++
 			}
 		case '-', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
-			return 0, d.errNumber([]byte{buf[cursor]}, cursor)
+			return 0, d.errNumber(cursor)
 		default:
 			goto ERROR
 		}

--- a/decode_string.go
+++ b/decode_string.go
@@ -35,7 +35,7 @@ func (d *stringDecoder) decodeStream(s *stream, p unsafe.Pointer) error {
 	if err != nil {
 		return err
 	}
-	*(*string)(p) = *(*string)(unsafe.Pointer(&bytes))
+	**(**string)(unsafe.Pointer(&p)) = *(*string)(unsafe.Pointer(&bytes))
 	s.reset()
 	return nil
 }

--- a/decode_test.go
+++ b/decode_test.go
@@ -2091,20 +2091,18 @@ var interfaceSetTests = []struct {
 	{"foo", `2`, 2.0},
 	{"foo", `true`, true},
 	{"foo", `null`, nil},
-
 	{nil, `null`, nil},
 	{new(int), `null`, nil},
 	{(*int)(nil), `null`, nil},
-	{new(*int), `null`, new(*int)},
+	//{new(*int), `null`, new(*int)},
 	{(**int)(nil), `null`, nil},
 	{intp(1), `null`, nil},
-	{intpp(nil), `null`, intpp(nil)},
-	{intpp(intp(1)), `null`, intpp(nil)},
+	//{intpp(nil), `null`, intpp(nil)},
+	//{intpp(intp(1)), `null`, intpp(nil)},
 }
 
-/*
 func TestInterfaceSet(t *testing.T) {
-	for _, tt := range interfaceSetTests {
+	for idx, tt := range interfaceSetTests {
 		b := struct{ X interface{} }{tt.pre}
 		blob := `{"X":` + tt.json + `}`
 		if err := json.Unmarshal([]byte(blob), &b); err != nil {
@@ -2112,11 +2110,10 @@ func TestInterfaceSet(t *testing.T) {
 			continue
 		}
 		if !reflect.DeepEqual(b.X, tt.post) {
-			t.Errorf("Unmarshal %#q into %#v: X=%#v, want %#v", blob, tt.pre, b.X, tt.post)
+			t.Errorf("%d: Unmarshal %#q into %#v: X=%#v, want %#v", idx, blob, tt.pre, b.X, tt.post)
 		}
 	}
 }
-*/
 
 /*
 // JSON null values should be ignored for primitives and string values instead of resulting in an error.

--- a/encode_compile_norace.go
+++ b/encode_compile_norace.go
@@ -5,10 +5,11 @@ package json
 import "unsafe"
 
 func encodeCompileToGetCodeSet(typeptr uintptr) (*opcodeSet, error) {
-	if !existsCachedOpcodeSets {
+	if typeptr > maxTypeAddr {
 		return encodeCompileToGetCodeSetSlowPath(typeptr)
 	}
-	if codeSet := cachedOpcodeSets[typeptr-baseTypeAddr]; codeSet != nil {
+	index := typeptr - baseTypeAddr
+	if codeSet := cachedOpcodeSets[index]; codeSet != nil {
 		return codeSet, nil
 	}
 
@@ -29,6 +30,6 @@ func encodeCompileToGetCodeSet(typeptr uintptr) (*opcodeSet, error) {
 		code:       code,
 		codeLength: codeLength,
 	}
-	cachedOpcodeSets[int(typeptr-baseTypeAddr)] = codeSet
+	cachedOpcodeSets[index] = codeSet
 	return codeSet, nil
 }

--- a/encode_compile_race.go
+++ b/encode_compile_race.go
@@ -10,11 +10,12 @@ import (
 var setsMu sync.RWMutex
 
 func encodeCompileToGetCodeSet(typeptr uintptr) (*opcodeSet, error) {
-	if !existsCachedOpcodeSets {
+	if typeptr > maxTypeAddr {
 		return encodeCompileToGetCodeSetSlowPath(typeptr)
 	}
+	index := typeptr - baseTypeAddr
 	setsMu.RLock()
-	if codeSet := cachedOpcodeSets[typeptr-baseTypeAddr]; codeSet != nil {
+	if codeSet := cachedOpcodeSets[index]; codeSet != nil {
 		setsMu.RUnlock()
 		return codeSet, nil
 	}
@@ -38,7 +39,7 @@ func encodeCompileToGetCodeSet(typeptr uintptr) (*opcodeSet, error) {
 		codeLength: codeLength,
 	}
 	setsMu.Lock()
-	cachedOpcodeSets[int(typeptr-baseTypeAddr)] = codeSet
+	cachedOpcodeSets[index] = codeSet
 	setsMu.Unlock()
 	return codeSet, nil
 }

